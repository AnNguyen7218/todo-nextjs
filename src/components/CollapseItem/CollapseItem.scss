@import 'config.gutters';
@import 'config.colors';
@import 'config.media';

.container {
  margin-bottom: gutter(16);
  padding-bottom: gutter(16);
  border-bottom: 0.05rem solid $black;
  cursor: pointer;

  @include break(xl) {
    margin-bottom: gutter(24);
  }
}

.header-container {
  position: relative;
}

.title {
  width: 90%;
}

.icon {
  position: absolute;
  width: 2.4rem;
  height: 2.4rem;
  right: 0.6rem;
  top: 0;
  color: $black;
}

.content {
  overflow: hidden;
}
