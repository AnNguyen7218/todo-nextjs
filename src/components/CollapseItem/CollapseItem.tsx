import React, { useState, useEffect } from 'react';
import { motion, Variants } from 'framer-motion';
import { XL } from 'constants/media';
import Expanse from 'assets/icons/ic-add.svg';
import Collapse from 'assets/icons/ic-close.svg';
import css from './CollapseItem.scss';

type Props = {
  item: FaqCollapseItemType;
};

const MARGIN_MOBILE = 8;
const MARGIN_WEB = 20;

const variants: Variants = {
  expanded: contentMargin => ({
    height: '100%',
    transition: { duration: 0.5, ease: 'easeOut' },
    marginTop: contentMargin,
  }),
  collapsed: {
    height: 0,
    transition: { duration: 0.5, ease: 'easeIn' },
    marginTop: 0,
  },
};

const CollapseItem: React.FC<Props> = ({ item: { title, content } }) => {
  const [isActive, setActive] = useState(false);
  const [contentMargin, setContentMargin] = useState(0);
  const Icon = isActive ? Collapse : Expanse;

  const onItemClick = () => {
    setActive(!isActive);
  };

  const onWindowResize = () => {
    const currentWidth = window.innerWidth;
    currentWidth < XL
      ? setContentMargin(MARGIN_MOBILE)
      : setContentMargin(MARGIN_WEB);
  };

  useEffect(() => {
    onWindowResize();
    window.addEventListener('resize', onWindowResize);
  });

  return (
    <div onClick={onItemClick} className={css['container']}>
      <div className={css['header-container']}>
        <h3 className={css['title']}>{title}</h3>
        <Icon className={css['icon']} />
      </div>
      <motion.div
        custom={contentMargin}
        className={css['content']}
        initial={'collapse'}
        variants={variants}
        animate={isActive ? 'expanded' : 'collapsed'}
      >
        <p>{content}</p>
      </motion.div>
    </div>
  );
};

export default CollapseItem;
