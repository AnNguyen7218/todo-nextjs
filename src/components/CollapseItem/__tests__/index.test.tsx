import React from 'react';
import { render, cleanup } from '@testing-library/react';
import CollapseItem from '../CollapseItem';

afterEach(cleanup);

const item = {
  title: 'title',
  content: 'content',
};

describe('CollapseItem Component', () => {
  test('CollapseItem should match snapshot', () => {
    const { container } = render(<CollapseItem item={item} />);
    expect(container).toMatchSnapshot();
  });
});
