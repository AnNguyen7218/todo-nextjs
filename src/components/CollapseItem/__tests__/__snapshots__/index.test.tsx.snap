// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapseItem Component CollapseItem should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="header-container"
    >
      <h3
        class="title"
      >
        title
      </h3>
      <ic-add.svg
        classname="icon"
      />
    </div>
    <div
      class="content"
      style="height: 0px; margin-top: 0px;"
    >
      <p>
        content
      </p>
    </div>
  </div>
</div>
`;
